### PR TITLE
Use CopyOnWriteArrayList for killAllListeners

### PR DESCRIPTION
**Pull request by mfussenegger**
## Summary of the changes / Why this improves CrateDB

The killAllListeners are only written to on node start-up and are
afterwards effectively read-only (on each kill request), so a
CopyOnWriteArrayList suites the workload better.

## Checklist

 - [x] User relevant changes are recorded in ``CHANGES.txt``
 - [x] Touched code is covered by tests
 - [x] Documentation has been updated if necessary
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

**Comment by mfussenegger** [comment]
retest this please

File: sql/src/main/java/io/crate/execution/jobs/TasksService.java [hunk 1]
```diff
@@ -58,7 +58,7 @@
     private final ConcurrentMap<UUID, RootTask> activeTasks =
         ConcurrentCollections.newConcurrentMapWithAggressiveConcurrency();
 
-    private final List<KillAllListener> killAllListeners = Collections.synchronizedList(new ArrayList<>());
+    private final List<KillAllListener> killAllListeners = new CopyOnWriteArrayList<>();
```
Comment (andreidan) [inline sql/src/main/java/io/crate/execution/jobs/TasksService.java]: hm, shall we do this? with synchronizedList we're in danger of ConcurrentModificationException while we iterated over the listeners so we do need to fix this, but is COWAL the answer? (we mostly write in this list and COWAL is way slower and creates garbage) IMO we should keep synchronizedList and create a manual copy only when we do iterate over it and kill all the listeners in this "copy list"

Comment (andreidan) [reply sql/src/main/java/io/crate/execution/jobs/TasksService.java]: thinking about it some more, the size of this list is usually relatively small so maybe the COWAL overhead is not significant

Comment (mfussenegger) [reply sql/src/main/java/io/crate/execution/jobs/TasksService.java]: See the description in the commit message and PR description. The writes only happen once on node start-up (and I think currently it is just 1 write)

Comment (andreidan) [reply sql/src/main/java/io/crate/execution/jobs/TasksService.java]: yep, makes sense (didn't see the @Singleton/Insert on TransportShardUpsertAction), good catch (lock free reads ❤️ )

**Pull request merged**